File: scripts/translate.js
const fs = require('fs');
const path = require('path');
const https = require('https');

const DEEPL_API_URL = 'https://api-free.deepl.com/v2/translate';

function loadEnv() {
  const envPath = path.join(__dirname, '..', '.env');
  if (fs.existsSync(envPath)) {
    const lines = fs.readFileSync(envPath, 'utf-8').split('\n');
    for (const line of lines) {
      const trimmed = line.trim();
      if (trimmed && !trimmed.startsWith('#')) {
        const eqIdx = trimmed.indexOf('=');
        if (eqIdx > 0) {
          const key = trimmed.substring(0, eqIdx).trim();
          const value = trimmed.substring(eqIdx + 1).trim();
          if (!process.env[key]) {
            process.env[key] = value;
          }
        }
      }
    }
  }
}

function translateText(text, apiKey, retries = 3) {
  return new Promise((resolve, reject) => {
    if (!text || text.trim().length === 0) return resolve(text);

    const postData = JSON.stringify({
      text: [text],
      target_lang: 'ES',
      source_lang: 'EN',
      tag_handling: 'xml'
    });

    const url = new URL(DEEPL_API_URL);
    const options = {
      hostname: url.hostname,
      path: url.pathname,
      method: 'POST',
      headers: {
        'Authorization': `DeepL-Auth-Key ${apiKey}`,
        'Content-Type': 'application/json',
        'User-Agent': 'PulseRoots-Translator/1.0'
      }
    };

    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) {
          try {
            const result = JSON.parse(data);
            resolve(result.translations[0].text);
          } catch (e) {
            reject(new Error(`Parse error: ${e.message}`));
          }
        } else if (res.statusCode === 429 && retries > 0) {
          console.log(`  Rate limited, retrying in 2s... (${retries} retries left)`);
          setTimeout(() => {
            translateText(text, apiKey, retries - 1).then(resolve).catch(reject);
          }, 2000);
        } else if (res.statusCode === 456 && retries > 0) {
          console.log(`  Quota exceeded, retrying in 10s... (${retries} retries left)`);
          setTimeout(() => {
            translateText(text, apiKey, retries - 1).then(resolve).catch(reject);
          }, 10000);
        } else {
          reject(new Error(`DeepL API error ${res.statusCode}: ${data}`));
        }
      });
    });

    req.on('error', reject);
    req.write(postData);
    req.end();
  });
}

async function translateTree(items, apiKey, depth = 0) {
  const indent = '  '.repeat(depth);
  let translatedCount = 0;

  for (let i = 0; i < items.length; i++) {
    const item = items[i];
    const name = item.name || item.style || 'unknown';

    if (item.description) {
      process.stdout.write(`${indent}Translating "${name}"... `);
      try {
        const translated = await translateText(item.description, apiKey);
        item.description = translated;
        process.stdout.write('✓\n');
        translatedCount++;
      } catch (err) {
        process.stdout.write(`✗ ${err.message}\n`);
      }
      await new Promise(r => setTimeout(r, 200));
    }

    if (item.substyles && item.substyles.length > 0) {
      const childCount = await translateTree(item.substyles, apiKey, depth + 1);
      translatedCount += childCount;
    }
  }

  return translatedCount;
}

async function translateHistory(historyItems, apiKey) {
  let translatedCount = 0;

  for (let i = 0; i < historyItems.length; i++) {
    const item = historyItems[i];

    if (item.fact) {
      process.stdout.write(`  Translating history entry ${i + 1}/${historyItems.length} (${item.date})... `);
      try {
        const translated = await translateText(item.fact, apiKey);
        item.fact = translated;
        process.stdout.write('✓\n');
        translatedCount++;
      } catch (err) {
        process.stdout.write(`✗ ${err.message}\n`);
      }
      await new Promise(r => setTimeout(r, 150));
    }
  }

  return translatedCount;
}

async function main() {
  loadEnv();

  const apiKey = process.env.DEEPL_API_KEY;
  if (!apiKey) {
    console.error('Error: DEEPL_API_KEY not found in .env or environment.');
    console.error('');
    console.error('To get a free DeepL API key:');
    console.error('  1. Go to https://www.deepl.com/pro#developer');
    console.error('  2. Sign up for the free DeepL API plan (500K chars/month)');
    console.error('  3. Add to .env: DEEPL_API_KEY=your_key_here');
    console.error('');
    console.error('Or set the environment variable:');
    console.error('  $env:DEEPL_API_KEY="your_key_here"');
    process.exit(1);
  }

  const dataDir = path.join(__dirname, '..', 'data');

  // --- Translate genre descriptions ---
  console.log('\n--- Translating genre descriptions ---');
  const genresPath = path.join(dataDir, 'pulseroots.genres.json');
  const genresOutputPath = path.join(dataDir, 'pulseroots.genres.es.json');

  const genresData = JSON.parse(fs.readFileSync(genresPath, 'utf-8'));
  const genreCount = await translateTree(genresData, apiKey);
  fs.writeFileSync(genresOutputPath, JSON.stringify(genresData, null, 4), 'utf-8');
  console.log(`\n✓ Genre translations complete: ${genreCount} descriptions translated`);
  console.log(`  Output: ${genresOutputPath}`);

  // --- Translate music history ---
  console.log('\n--- Translating music history ---');
  const historyPath = path.join(dataDir, 'music_history.json');
  const historyOutputPath = path.join(dataDir, 'music_history.es.json');

  const historyData = JSON.parse(fs.readFileSync(historyPath, 'utf-8'));
  const historyCount = await translateHistory(historyData, apiKey);
  fs.writeFileSync(historyOutputPath, JSON.stringify(historyData, null, 2), 'utf-8');
  console.log(`\n✓ History translations complete: ${historyCount} entries translated`);
  console.log(`  Output: ${historyOutputPath}`);

  console.log('\n✅ All translations complete!');
  console.log('Next step: manually review music terminology in the Spanish files.');
  console.log('Then run: node build.js --lang es');
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: scripts/translate.py
import json
import os
import sys
import time
import re
from deep_translator import GoogleTranslator

sys.stdout.reconfigure(encoding='utf-8')

DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'data')

GENRES_INPUT = os.path.join(DATA_DIR, 'pulseroots.genres.json')
GENRES_OUTPUT = os.path.join(DATA_DIR, 'pulseroots.genres.es.json')
HISTORY_INPUT = os.path.join(DATA_DIR, 'music_history.json')
HISTORY_OUTPUT = os.path.join(DATA_DIR, 'music_history.es.json')

translator = GoogleTranslator(source='en', target='es')

def untag_text(text):
    text = re.sub(r'<[^>]+>', '', text)
    text = text.replace('&amp;', '&').replace('&lt;', '<').replace('&gt;', '>').replace('&quot;', '"')
    return text

def translate_text(text, retries=3):
    if not text or not text.strip():
        return text
    for attempt in range(retries):
        try:
            result = translator.translate(text)
            time.sleep(0.3)
            return result
        except Exception as e:
            print(f"  Retry {attempt+1}/{retries}: {e}")
            time.sleep(2)
    print(f"  FAILED after {retries} retries")
    return text

def translate_tree(items, depth=0):
    indent = '  ' * depth
    count = 0
    for i, item in enumerate(items):
        name = item.get('name') or item.get('style') or 'unknown'

        if 'description' in item and item['description']:
            desc = item['description']
            if len(desc) > 5:
                print(f"{indent}[{i+1}/{len(items)}] Translating \"{name}\"...", end=' ')
                sys.stdout.flush()
                try:
                    translated = translate_text(desc)
                    item['description'] = translated
                    print('OK')
                    count += 1
                except Exception as e:
                    print(f'ERROR: {e}')

        if 'substyles' in item and item['substyles']:
            c = translate_tree(item['substyles'], depth + 1)
            count += c

    return count

def translate_history(items):
    total = len(items)
    count = 0
    for i, item in enumerate(items):
        if 'fact' in item and item['fact']:
            print(f"  History [{i+1}/{total}] ({item.get('date', '')})...", end=' ')
            sys.stdout.flush()
            try:
                translated = translate_text(item['fact'])
                item['fact'] = translated
                print('OK')
                count += 1
            except Exception as e:
                print(f'ERROR: {e}')
    return count

def main():
    print("=" * 60)
    print("PulseRoots Translation Script (EN -> ES)")
    print("=" * 60)

    # --- Translate genres ---
    print("\n--- Translating genre descriptions ---")
    with open(GENRES_INPUT, 'r', encoding='utf-8') as f:
        genres_data = json.load(f)
    print(f"Loaded {len(genres_data)} top-level genres")

    genre_count = translate_tree(genres_data)
    print(f"\nTranslated {genre_count} genre descriptions")

    with open(GENRES_OUTPUT, 'w', encoding='utf-8') as f:
        json.dump(genres_data, f, ensure_ascii=False, indent=4)
    print(f"Written to: {GENRES_OUTPUT}")

    # --- Translate history ---
    print("\n--- Translating music history ---")
    with open(HISTORY_INPUT, 'r', encoding='utf-8') as f:
        history_data = json.load(f)
    print(f"Loaded {len(history_data)} history entries")

    history_count = translate_history(history_data)
    print(f"\nTranslated {history_count} history entries")

    with open(HISTORY_OUTPUT, 'w', encoding='utf-8') as f:
        json.dump(history_data, f, ensure_ascii=False, indent=2)
    print(f"Written to: {HISTORY_OUTPUT}")

    print("\n" + "=" * 60)
    print("Translation complete!")
    print(f"  Genres: {genre_count} descriptions")
    print(f"  History: {history_count} entries")
    print("=" * 60)

if __name__ == '__main__':
    main()
